### PR TITLE
Add factory function derivatives

### PR DESCRIPTION
**Pull request by drisspg**
Adding derivatives for factory functions, this issue is used for tracking: #79044

### CODE REVIEW

File: aten/src/ATen/native/nested/NestedTensorMath.cpp [hunk 1]
```diff
@@ -372,7 +372,7 @@ Tensor NestedTensor_to_padded_tensor_generic(
 
   if (sizes.numel() == 0 || sizes.dim() == 0) {
     TORCH_INTERNAL_ASSERT_DEBUG_ONLY(nt.get_buffer().numel() == 0);
-    return nt.get_buffer();
+    return nt.get_buffer().clone();
```
Comment (drisspg) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:375]: use_count >1 for empty tensors fix

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 2]
```diff
@@ -28,6 +28,7 @@
 #include <ciso646>
 #include <functional>
 #include <numeric>
+#include "ATen/core/ATen_fwd.h"
```
Comment (drisspg) [inline torch/csrc/autograd/FunctionsManual.cpp:31]: no no

**Comment by drisspg** [comment]
`memory: you tried to allocate 14744769207296 bytes. Error code 12 (Cannot allocate memory)`

During autograd trying to allocate 14.75 TB of data... hmmm

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 3]
```diff
@@ -338,6 +338,22 @@ Tensor norm_jvp(
   return norm_jvp(self_p, self_t, p_, norm, {}, true);
 }
 
+Tensor _nested_from_padded_backward(
+    const Tensor& grad,
+    const Tensor& input,
+    const bool do_transform_0213) {
```
Comment (soulitzer) [inline torch/csrc/autograd/FunctionsManual.cpp:344]: nit: bool doesn't need to be const

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 4]
```diff
@@ -338,6 +338,22 @@ Tensor norm_jvp(
...
+  if (do_transform_0213) {
+    auto new_sizes = IntArrayRef{
+        input.size(0), input.size(2), (input.size(1) * input.size(3))};
+    auto out = grad.to_padded_tensor(0, new_sizes);
+    auto expand_last_dim_size =
+        IntArrayRef{input.size(0), input.size(2), input.size(1), input.size(3)};
+    out = out.view(expand_last_dim_size).permute({0, 2, 1, 3});
+    return out.reshape(input.sizes());
```
Comment (soulitzer) [inline torch/csrc/autograd/FunctionsManual.cpp:352]: why do we need this reshape? shouldn't the shape already be correct after the permute?

Comment (drisspg) [reply torch/csrc/autograd/FunctionsManual.cpp:352]: yeah, good catch

**Review by cpuhrsch — APPROVED** [review]
Great!

File: test/test_nestedtensor.py [hunk 5]
```diff
@@ -596,6 +596,77 @@ def test_backward_for_add_op(self):
...
+            out = nt.to_padded_tensor(padding_val)
+            grad_output = torch.ones(out.shape)
+            out.backward(grad_output)
+
+            self.nt_equal(nt.grad, torch.nested_tensor([torch.ones(1, 2), torch.ones(7, 8)]))
+
+    def test_nested_tensor_from_mask_and_to_padded(self):
+        N, L, D = 2, 4, 4
+        mask = torch.ones(N, L)
+        for i in range(N):
```
Comment (soulitzer) [inline test/test_nestedtensor.py:614]: nit: Do you mean `for i in range(1, N):` to avoid the extra `mask[0, :]`

File: test/test_nestedtensor.py [hunk 6]
```diff
@@ -596,6 +596,77 @@ def test_backward_for_add_op(self):
...
+            mask[i, end:] = 0
+
+        mask[0, :] = 1
+        mask = mask.bool()
+
+        data = torch.randn(N, L, D, requires_grad=True, dtype=torch.float64)
+
+        def grad_test_func(inpt):
+            nt = torch._nested_tensor_from_mask(inpt, mask)
+            # This implictily tests to_padded_tensor grads
```
Comment (soulitzer) [inline test/test_nestedtensor.py:625]: nit: implicitly

same with below

File: test/test_nestedtensor.py [hunk 7]
```diff
@@ -596,6 +596,77 @@ def test_backward_for_add_op(self):
...
+
+    def test_nested_tensor_from_mask_and_to_padded(self):
+        N, L, D = 2, 4, 4
+        mask = torch.ones(N, L)
+        for i in range(N):
+            end = torch.randint(1, L - 1, (1,))
+            mask[i, end:] = 0
+
+        mask[0, :] = 1
+        mask = mask.bool()
```
Comment (soulitzer) [inline test/test_nestedtensor.py:619]: Question (no action needed): is the semantics of broadcasting for the masking here is that the left-most dimensions need to be the same?

**Review by soulitzer — APPROVED** [review]
Cool!

File: test/test_nestedtensor.py (hunk 5 above)
Comment (drisspg) [reply test/test_nestedtensor.py:614]: yeah good catch, yeah at least one row should be as long as L since if it were smaller the padded tensor would not need to be all the way padded to L

File: test/test_nestedtensor.py (hunk 7 above)
Comment (drisspg) [reply test/test_nestedtensor.py:619]: Yeah correct here in the code: https://github.com/pytorch/pytorch/blob/f5eb05f10784c2c0b620c9b19ef8f6eb5e644741/aten/src/ATen/native/nested/NestedTensorMath.cpp#L170
